File: examples/39-benchmarking-without-lies/cpp/src/main.cpp
// benchlab (C++23) — proper benchmarking of the chatterd frame codec (ch21).
//
// Usage:
//   app --op encode|decode|roundtrip --iters N [--warmup W]   # the real thing
//   app --lie [--op encode|decode|roundtrip]                  # the anti-pattern
//
// The "real thing" times each iteration individually with a monotonic clock,
// discards a warmup phase, and reports min/median/p99/max plus a
// coordinated-omission-corrected p99 (see co_correct below). `--lie` times a
// single unwarmed call with wall-clock and reports one number — exactly the
// benchmark the rest of this example exists to discredit.
//
// Wire format (canonical chatterd chat frame, introduced ch21):
//   [ magic 0x43 0x48 ][ version 0x01 ][ type u8 ][ length u16 be ][ payload ]
// This file re-implements only encode_frame/decode_frame — the codec under
// test — not the daemon; see ch21/22/27 for the networked version.

#include <algorithm>
#include <cerrno>
#include <chrono>
#include <cstdint>
#include <cstdlib>
#include <expected>
#include <print>
#include <string>
#include <string_view>
#include <vector>

namespace {

// ---------------------------------------------------------------------------
// Frame codec under test — canonical chatterd chat frame (see file header).
// ---------------------------------------------------------------------------
constexpr std::size_t kHeaderLen = 6;
constexpr std::uint8_t kMagic0 = 0x43;
constexpr std::uint8_t kMagic1 = 0x48;
constexpr std::uint8_t kVersion = 0x01;
constexpr std::uint8_t kTypeDeliver = 3;

struct DecodedFrame {
    std::uint8_t type;
    std::string body;
};

std::string encode_frame(std::uint8_t type, std::string_view body) {
    std::string out;
    out.reserve(kHeaderLen + body.size());
    out.push_back(static_cast<char>(kMagic0));
    out.push_back(static_cast<char>(kMagic1));
    out.push_back(static_cast<char>(kVersion));
    out.push_back(static_cast<char>(type));
    auto len = static_cast<std::uint16_t>(body.size());
    out.push_back(static_cast<char>((len >> 8) & 0xff));
    out.push_back(static_cast<char>(len & 0xff));
    out.append(body);
    return out;
}

// Decodes exactly one frame that fills `buf` completely — no partial-read
// reassembly here, that lives in the networked chatterd (ch21/22/27); this
// harness only measures the codec's per-call cost.
[[nodiscard]] std::expected<DecodedFrame, std::string> decode_frame(std::string_view buf) {
    if (buf.size() < kHeaderLen) {
        return std::unexpected("frame shorter than header");
    }
    if (static_cast<std::uint8_t>(buf[0]) != kMagic0 || static_cast<std::uint8_t>(buf[1]) != kMagic1) {
        return std::unexpected("bad magic");
    }
    if (static_cast<std::uint8_t>(buf[2]) != kVersion) {
        return std::unexpected("bad version");
    }
    std::uint8_t type = static_cast<std::uint8_t>(buf[3]);
    std::uint16_t len = static_cast<std::uint16_t>((static_cast<std::uint8_t>(buf[4]) << 8) |
                                                    static_cast<std::uint8_t>(buf[5]));
    if (buf.size() != kHeaderLen + len) {
        return std::unexpected("length mismatch");
    }
    return DecodedFrame{type, std::string(buf.substr(kHeaderLen))};
}

// ---------------------------------------------------------------------------
// The fixed workload: a DELIVER frame carrying a realistic chat line. Every
// language in this example encodes/decodes the identical bytes.
// ---------------------------------------------------------------------------
constexpr std::string_view kNick = "alice";
constexpr std::string_view kText = "the quick brown fox jumps over the lazy dog, three times, for benchlab";

std::string delivery_body() {
    std::string body;
    body.reserve(kNick.size() + 1 + kText.size());
    body.append(kNick);
    body.push_back('\0');
    body.append(kText);
    return body;
}

enum class Op { Encode, Decode, Roundtrip };

std::expected<Op, std::string> parse_op(std::string_view s) {
    if (s == "encode") return Op::Encode;
    if (s == "decode") return Op::Decode;
    if (s == "roundtrip") return Op::Roundtrip;
    return std::unexpected("unknown --op '" + std::string(s) + "'");
}

std::string_view op_name(Op op) {
    switch (op) {
        case Op::Encode: return "encode";
        case Op::Decode: return "decode";
        case Op::Roundtrip: return "roundtrip";
    }
    return "?";
}

// One iteration of the requested op against the fixed workload. Returns a
// small scalar derived from the result so the caller can fold it into a
// checksum — this is what keeps the optimizer from proving the call's result
// is unused and deleting it.
std::uint64_t run_once(Op op, const std::string& body, const std::string& prebuilt_frame) {
    switch (op) {
        case Op::Encode: {
            std::string frame = encode_frame(kTypeDeliver, body);
            return frame.empty() ? 0 : static_cast<std::uint8_t>(frame.back());
        }
        case Op::Decode: {
            auto decoded = decode_frame(prebuilt_frame);
            if (!decoded) {
                std::println(stderr, "benchlab: codec bug: decode failed: {}", decoded.error());
                std::exit(1);
            }
            return decoded->body.size();
        }
        case Op::Roundtrip: {
            std::string frame = encode_frame(kTypeDeliver, body);
            auto decoded = decode_frame(frame);
            if (!decoded) {
                std::println(stderr, "benchlab: codec bug: decode failed: {}", decoded.error());
                std::exit(1);
            }
            return decoded->body.size();
        }
    }
    return 0;
}

using Clock = std::chrono::steady_clock;

std::int64_t ns_between(Clock::time_point a, Clock::time_point b) {
    return std::chrono::duration_cast<std::chrono::nanoseconds>(b - a).count();
}

std::int64_t percentile(const std::vector<std::int64_t>& sorted, double p) {
    if (sorted.empty()) return 0;
    std::size_t idx = static_cast<std::size_t>(p * static_cast<double>(sorted.size() - 1));
    return sorted[idx];
}

// Coordinated-omission correction (HdrHistogram's recordValueWithExpectedInterval):
// this harness is a closed loop — iteration N+1 never starts until N returns —
// so a stall (a page fault, a context switch, a scheduling hiccup) is
// recorded as a single slow sample. A fixed-rate caller would instead have
// had many requests queue up behind that stall, each experiencing a similar
// multiple-of-the-target-interval delay. This backfills those missing
// virtual samples so the tail reflects what a real caller would have seen,
// not just what one lucky/unlucky iteration measured.
std::vector<std::int64_t> co_correct(const std::vector<std::int64_t>& raw,
                                      std::int64_t expected_interval_ns,
                                      std::size_t cap) {
    std::vector<std::int64_t> out;
    out.reserve(raw.size());
    for (std::int64_t v : raw) {
        out.push_back(v);
        if (expected_interval_ns <= 0 || v <= expected_interval_ns) continue;
        std::int64_t missing = v - expected_interval_ns;
        while (missing >= expected_interval_ns && out.size() < cap) {
            out.push_back(missing);
            missing -= expected_interval_ns;
        }
        if (out.size() >= cap) break;
    }
    return out;
}

void usage_and_exit() {
    std::println(stderr, "usage: benchlab --op encode|decode|roundtrip --iters N [--warmup W]");
    std::println(stderr, "       benchlab --lie [--op encode|decode|roundtrip]");
    std::exit(2);
}

std::expected<long long, std::string> parse_u64(std::string_view s) {
    if (s.empty()) return std::unexpected("empty number");
    long long v = 0;
    for (char c : s) {
        if (c < '0' || c > '9') return std::unexpected("not a number: " + std::string(s));
        v = v * 10 + (c - '0');
    }
    return v;
}

} // namespace

int main(int argc, char** argv) {
    bool lie = false;
    std::string op_str;
    long long iters = -1;
    long long warmup = 1000;

    for (int i = 1; i < argc; ++i) {
        std::string_view a = argv[i];
        if (a == "--lie") {
            lie = true;
        } else if (a == "--op" && i + 1 < argc) {
            op_str = argv[++i];
        } else if (a == "--iters" && i + 1 < argc) {
            auto v = parse_u64(argv[++i]);
            if (!v) usage_and_exit();
            iters = *v;
        } else if (a == "--warmup" && i + 1 < argc) {
            auto v = parse_u64(argv[++i]);
            if (!v) usage_and_exit();
            warmup = *v;
        } else {
            usage_and_exit();
        }
    }

    if (!lie && op_str.empty()) usage_and_exit();
    Op op = Op::Roundtrip;
    if (!op_str.empty()) {
        auto parsed = parse_op(op_str);
        if (!parsed) usage_and_exit();
        op = *parsed;
    }

    const std::string body = delivery_body();
    const std::string prebuilt = encode_frame(kTypeDeliver, body);

    if (lie) {
        // The anti-pattern: no warmup, one call, wall-clock, done. This is
        // exactly the benchmark this example's README argues never to trust.
        auto t0 = Clock::now();
        std::uint64_t sink = run_once(op, body, prebuilt);
        auto t1 = Clock::now();
        std::println("benchlab: lie op={} (no warmup, single wall-clock sample, ignores variance)", op_name(op));
        std::println("benchlab: lie elapsed_ns={} sink={}", ns_between(t0, t1), sink);
        return 0;
    }

    if (iters < 2 || warmup < 0 || iters <= warmup) usage_and_exit();

    // Warmup: run the op without timing it, letting allocators/caches/branch
    // predictors reach steady state before any sample is recorded.
    std::uint64_t checksum = 0;
    for (long long i = 0; i < warmup; ++i) {
        checksum += run_once(op, body, prebuilt);
    }

    std::vector<std::int64_t> raw;
    raw.reserve(static_cast<std::size_t>(iters));
    for (long long i = 0; i < iters; ++i) {
        auto t0 = Clock::now();
        checksum += run_once(op, body, prebuilt);
        auto t1 = Clock::now();
        raw.push_back(ns_between(t0, t1));
    }

    std::vector<std::int64_t> sorted = raw;
    std::sort(sorted.begin(), sorted.end());
    std::int64_t min_ns = sorted.front();
    std::int64_t median_ns = percentile(sorted, 0.5);
    std::int64_t p99_ns = percentile(sorted, 0.99);
    std::int64_t max_ns = sorted.back();

    std::int64_t expected_interval_ns = median_ns > 0 ? median_ns : 1;
    constexpr std::size_t kCoCap = 5'000'000;
    std::vector<std::int64_t> corrected = co_correct(raw, expected_interval_ns, kCoCap);
    std::sort(corrected.begin(), corrected.end());
    std::int64_t co_p99_ns = percentile(corrected, 0.99);

    std::println("benchlab: op={} iters={} warmup={}", op_name(op), iters, warmup);
    std::println("benchlab: n={} min_ns={} median_ns={} p99_ns={} max_ns={}",
                 sorted.size(), min_ns, median_ns, p99_ns, max_ns);
    std::println("benchlab: co_p99_ns={} expected_interval_ns={} co_n={}",
                 co_p99_ns, expected_interval_ns, corrected.size());
    std::println("benchlab: checksum={:016x}", checksum);
    return 0;
}

File: examples/39-benchmarking-without-lies/rust/src/main.rs
// benchlab (Rust) — proper benchmarking of the chatterd frame codec (ch21).
//
// Usage:
//   app --op encode|decode|roundtrip --iters N [--warmup W]   # the real thing
//   app --lie [--op encode|decode|roundtrip]                  # the anti-pattern
//
// The "real thing" times each iteration individually with a monotonic clock,
// discards a warmup phase, and reports min/median/p99/max plus a
// coordinated-omission-corrected p99 (see co_correct below). `--lie` times a
// single unwarmed call with wall-clock and reports one number — exactly the
// benchmark the rest of this example exists to discredit.
//
// Wire format (canonical chatterd chat frame, introduced ch21):
//   [ magic 0x43 0x48 ][ version 0x01 ][ type u8 ][ length u16 be ][ payload ]
// This file re-implements only encode_frame/decode_frame — the codec under
// test — not the daemon; see ch21/22/27 for the networked version.

use std::process::ExitCode;
use std::time::Instant;

// ---------------------------------------------------------------------------
// Frame codec under test — canonical chatterd chat frame (see file header).
// ---------------------------------------------------------------------------
const HEADER_LEN: usize = 6;
const MAGIC0: u8 = 0x43;
const MAGIC1: u8 = 0x48;
const VERSION: u8 = 0x01;
const TYPE_DELIVER: u8 = 3;

fn encode_frame(typ: u8, body: &[u8]) -> Vec<u8> {
    let len = body.len() as u16;
    let mut out = Vec::with_capacity(HEADER_LEN + body.len());
    out.push(MAGIC0);
    out.push(MAGIC1);
    out.push(VERSION);
    out.push(typ);
    out.extend_from_slice(&len.to_be_bytes());
    out.extend_from_slice(body);
    out
}

// Decodes exactly one frame that fills `buf` completely — no partial-read
// reassembly here, that lives in the networked chatterd (ch21/22/27); this
// harness only measures the codec's per-call cost.
fn decode_frame(buf: &[u8]) -> Result<(u8, Vec<u8>), String> {
    if buf.len() < HEADER_LEN {
        return Err("frame shorter than header".to_string());
    }
    if buf[0] != MAGIC0 || buf[1] != MAGIC1 {
        return Err("bad magic".to_string());
    }
    if buf[2] != VERSION {
        return Err("bad version".to_string());
    }
    let typ = buf[3];
    let len = u16::from_be_bytes([buf[4], buf[5]]) as usize;
    if buf.len() != HEADER_LEN + len {
        return Err("length mismatch".to_string());
    }
    Ok((typ, buf[HEADER_LEN..].to_vec()))
}

// ---------------------------------------------------------------------------
// The fixed workload: a DELIVER frame carrying a realistic chat line. Every
// language in this example encodes/decodes the identical bytes.
// ---------------------------------------------------------------------------
const NICK: &str = "alice";
const TEXT: &str = "the quick brown fox jumps over the lazy dog, three times, for benchlab";

fn delivery_body() -> Vec<u8> {
    let mut body = Vec::with_capacity(NICK.len() + 1 + TEXT.len());
    body.extend_from_slice(NICK.as_bytes());
    body.push(0);
    body.extend_from_slice(TEXT.as_bytes());
    body
}

#[derive(Clone, Copy)]
enum Op {
    Encode,
    Decode,
    Roundtrip,
}

impl Op {
    fn parse(s: &str) -> Result<Op, String> {
        match s {
            "encode" => Ok(Op::Encode),
            "decode" => Ok(Op::Decode),
            "roundtrip" => Ok(Op::Roundtrip),
            other => Err(format!("unknown --op '{other}'")),
        }
    }

    fn name(self) -> &'static str {
        match self {
            Op::Encode => "encode",
            Op::Decode => "decode",
            Op::Roundtrip => "roundtrip",
        }
    }
}

// One iteration of the requested op against the fixed workload. Returns a
// small scalar derived from the result so the caller can fold it into a
// checksum — this is what keeps the optimizer from proving the call's result
// is unused and deleting it. Propagates a codec bug (should never trigger on
// this fixed workload) via `?` up to main.
fn run_once(op: Op, body: &[u8], prebuilt_frame: &[u8]) -> Result<u64, String> {
    match op {
        Op::Encode => {
            let frame = encode_frame(TYPE_DELIVER, body);
            Ok(*frame.last().unwrap_or(&0) as u64)
        }
        Op::Decode => {
            let (_typ, payload) = decode_frame(prebuilt_frame)?;
            Ok(payload.len() as u64)
        }
        Op::Roundtrip => {
            let frame = encode_frame(TYPE_DELIVER, body);
            let (_typ, payload) = decode_frame(&frame)?;
            Ok(payload.len() as u64)
        }
    }
}

fn percentile(sorted: &[i64], p: f64) -> i64 {
    if sorted.is_empty() {
        return 0;
    }
    let idx = (p * (sorted.len() - 1) as f64) as usize;
    sorted[idx]
}

// Coordinated-omission correction (HdrHistogram's recordValueWithExpectedInterval):
// this harness is a closed loop — iteration N+1 never starts until N returns —
// so a stall (a page fault, a context switch, a scheduling hiccup) is
// recorded as a single slow sample. A fixed-rate caller would instead have
// had many requests queue up behind that stall, each experiencing a similar
// multiple-of-the-target-interval delay. This backfills those missing
// virtual samples so the tail reflects what a real caller would have seen,
// not just what one lucky/unlucky iteration measured.
fn co_correct(raw: &[i64], expected_interval_ns: i64, cap: usize) -> Vec<i64> {
    let mut out = Vec::with_capacity(raw.len());
    'outer: for &v in raw {
        out.push(v);
        if expected_interval_ns <= 0 || v <= expected_interval_ns {
            continue;
        }
        let mut missing = v - expected_interval_ns;
        while missing >= expected_interval_ns {
            if out.len() >= cap {
                break 'outer;
            }
            out.push(missing);
            missing -= expected_interval_ns;
        }
    }
    out
}

fn usage() -> ! {
    eprintln!("usage: benchlab --op encode|decode|roundtrip --iters N [--warmup W]");
    eprintln!("       benchlab --lie [--op encode|decode|roundtrip]");
    std::process::exit(2);
}

fn parse_u64(s: &str) -> Option<i64> {
    if s.is_empty() || !s.bytes().all(|b| b.is_ascii_digit()) {
        return None;
    }
    s.parse::<i64>().ok()
}

struct Args {
    lie: bool,
    op: Option<Op>,
    iters: i64,
    warmup: i64,
}

fn parse_args() -> Args {
    let argv: Vec<String> = std::env::args().collect();
    let mut lie = false;
    let mut op_str: Option<String> = None;
    let mut iters: i64 = -1;
    let mut warmup: i64 = 1000;

    let mut i = 1;
    while i < argv.len() {
        match argv[i].as_str() {
            "--lie" => lie = true,
            "--op" if i + 1 < argv.len() => {
                op_str = Some(argv[i + 1].clone());
                i += 1;
            }
            "--iters" if i + 1 < argv.len() => {
                iters = parse_u64(&argv[i + 1]).unwrap_or_else(|| usage());
                i += 1;
            }
            "--warmup" if i + 1 < argv.len() => {
                warmup = parse_u64(&argv[i + 1]).unwrap_or_else(|| usage());
                i += 1;
            }
            _ => usage(),
        }
        i += 1;
    }

    if !lie && op_str.is_none() {
        usage();
    }
    let op = op_str.map(|s| Op::parse(&s).unwrap_or_else(|_| usage()));

    Args { lie, op, iters, warmup }
}

fn main() -> ExitCode {
    let args = parse_args();
    let op = args.op.unwrap_or(Op::Roundtrip);
    let body = delivery_body();
    let prebuilt = encode_frame(TYPE_DELIVER, &body);

    if args.lie {
        // The anti-pattern: no warmup, one call, wall-clock, done. This is
        // exactly the benchmark this example's README argues never to trust.
        let t0 = Instant::now();
        let sink = match run_once(op, &body, &prebuilt) {
            Ok(v) => v,
            Err(e) => {
                eprintln!("benchlab: codec bug: {e}");
                return ExitCode::from(1);
            }
        };
        let elapsed_ns = t0.elapsed().as_nanos();
        println!(
            "benchlab: lie op={} (no warmup, single wall-clock sample, ignores variance)",
            op.name()
        );
        println!("benchlab: lie elapsed_ns={elapsed_ns} sink={sink}");
        return ExitCode::SUCCESS;
    }

    if args.iters < 2 || args.warmup < 0 || args.iters <= args.warmup {
        usage();
    }
    let iters = args.iters;
    let warmup = args.warmup;

    // Warmup: run the op without timing it, letting allocators/caches/branch
    // predictors reach steady state before any sample is recorded.
    let mut checksum: u64 = 0;
    for _ in 0..warmup {
        match run_once(op, &body, &prebuilt) {
            Ok(v) => checksum = checksum.wrapping_add(v),
            Err(e) => {
                eprintln!("benchlab: codec bug: {e}");
                return ExitCode::from(1);
            }
        }
    }

    let mut raw: Vec<i64> = Vec::with_capacity(iters as usize);
    for _ in 0..iters {
        let t0 = Instant::now();
        let r = run_once(op, &body, &prebuilt);
        let elapsed_ns = t0.elapsed().as_nanos() as i64;
        match r {
            Ok(v) => checksum = checksum.wrapping_add(v),
            Err(e) => {
                eprintln!("benchlab: codec bug: {e}");
                return ExitCode::from(1);
            }
        }
        raw.push(elapsed_ns);
    }

    let mut sorted = raw.clone();
    sorted.sort_unstable();
    let min_ns = sorted[0];
    let median_ns = percentile(&sorted, 0.5);
    let p99_ns = percentile(&sorted, 0.99);
    let max_ns = *sorted.last().unwrap();

    let expected_interval_ns = if median_ns > 0 { median_ns } else { 1 };
    const CO_CAP: usize = 5_000_000;
    let mut corrected = co_correct(&raw, expected_interval_ns, CO_CAP);
    corrected.sort_unstable();
    let co_p99_ns = percentile(&corrected, 0.99);

    println!("benchlab: op={} iters={} warmup={}", op.name(), iters, warmup);
    println!(
        "benchlab: n={} min_ns={} median_ns={} p99_ns={} max_ns={}",
        sorted.len(),
        min_ns,
        median_ns,
        p99_ns,
        max_ns
    );
    println!(
        "benchlab: co_p99_ns={} expected_interval_ns={} co_n={}",
        co_p99_ns,
        expected_interval_ns,
        corrected.len()
    );
    println!("benchlab: checksum={checksum:016x}");
    ExitCode::SUCCESS
}
